support-services/src/main/scala/com/gu/support/promotions/PromotionService.scala

```diff
@@ -25,12 +25,17 @@ class PromotionService(config: PromotionsConfig, maybeCollection: Option[Promoti
       case tooMany => Left(DuplicateCode(tooMany.mkString(", ")))
     }
 
-  def findPromotions(promoCodes: List[PromoCode]): List[PromotionWithCode] =
-    allWith6For6
-      .foldLeft(List.empty[PromotionWithCode]) { (acc, promotion) =>
-        val maybeCode = promoCodes.intersect(promotion.promoCodes.toList).headOption
-        maybeCode.map(code => acc :+ PromotionWithCode(code, promotion)).getOrElse(acc)
-      }
+  // promoCodes here is expected to be a small list of default promos plus one from the querystring
+  def findPromotions(promoCodes: List[PromoCode]): List[PromotionWithCode] = {
+    val promosByCode: Map[PromoCode, List[Promotion]] =
+      allWith6For6
+        .flatMap(promo => promo.promoCodes.map(code => code -> promo))
+        .groupMap(_._1) { case (promoCode, promo) => promo }
```

Comment:
`groupMap` is nice - the first function defines how to group the items, and the second defines the values in the new map.